sentry-actix/src/lib.rs

```diff
@@ -68,221 +71,267 @@
+        inner.capture_server_errors = val;
         self
     }
 }
 
-/// Reports certain failures to sentry.
-pub struct SentryMiddleware {
+struct Inner {
     hub: Option<Arc<Hub>>,
     emit_header: bool,
     capture_server_errors: bool,
 }
 
+/// Reports certain failures to Sentry.
+pub struct Sentry(Rc<Inner>);
+
 struct HubWrapper {
     hub: Arc<Hub>,
-    root_scope: RefCell<Option<ScopeGuard>>,
 }
 
-impl SentryMiddleware {
+impl Sentry {
     /// Creates a new sentry middleware.
-    pub fn new() -> SentryMiddleware {
-        SentryMiddleware {
+    pub fn new() -> Self {
+        Sentry(Rc::new(Inner {
             hub: None,
             emit_header: false,
             capture_server_errors: true,
-        }
+        }))
     }
 
     /// Creates a new middleware builder.
-    pub fn builder() -> SentryMiddlewareBuilder {
-        SentryMiddleware::new().into_builder()
+    pub fn builder() -> SentryBuilder {
+        Sentry::new().into_builder()
     }
 
     /// Converts the middleware into a builder.
-    pub fn into_builder(self) -> SentryMiddlewareBuilder {
-        SentryMiddlewareBuilder { middleware: self }
+    pub fn into_builder(self) -> SentryBuilder {
+        SentryBuilder { middleware: self }
     }
+}
 
-    fn new_hub(&self) -> Arc<Hub> {
-        Arc::new(Hub::new_from_top(Hub::main()))
+impl Default for Sentry {
+    fn default() -> Self {
+        Sentry::new()
     }
 }
 
-impl Default for SentryMiddleware {
-    fn default() -> Self {
-        SentryMiddleware::new()
+impl<S, B> Transform<S> for Sentry
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Transform = SentryMiddleware<S>;
+    type InitError = ();
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ok(SentryMiddleware {
+            service,
+            inner: self.0.clone(),
+        })
     }
 }
 
-fn extract_request<S: 'static>(
-    req: &HttpRequest<S>,
+/// The middleware for individual services.
+pub struct SentryMiddleware<S> {
+    service: S,
+    inner: Rc<Inner>,
+}
+
+impl<S, B> Service for SentryMiddleware<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: ServiceRequest) -> Self::Future {
+        let inner = self.inner.clone();
+        let hub = Arc::new(Hub::new_from_top(Hub::main()));
```

Comment:
I wonder if this has been broken all along. Why does the builder have options to set a specific Hub if we derive from the Main Hub anyway…

Reply:
That's my mistake, it should use the given Hub if specified and fall-back to the Main hub otherwise.

Is using `Hub::current` sufficient to get the hub or is it necessary to use `Hub::new_from_top`? See latest commit for context.

Reply:
You should still call `new_from_top`, otherwise concurrent futures will mess with each others scope, such as breadcrumbs.
Also, `add_event_processor` callbacks would multiply and leak for each request.